include/sys/dmu.h

```diff
@@ -532,6 +532,26 @@ void dmu_redact(objset_t *os, uint64_t object, uint64_t offset, uint64_t size,
 void dmu_write_policy(objset_t *os, dnode_t *dn, int level, int wp,
     struct zio_prop *zp);
 
+/*
+ * DB_RF_* are to be used for dbuf_read() or in limited other cases.
+ */
+typedef enum dmu_flags {
+	DB_RF_MUST_SUCCEED	= 0,	  /* Suspend on I/O errors. */
+	DB_RF_CANFAIL		= 1 << 0, /* Return on I/O errors. */
+	DB_RF_HAVESTRUCT	= 1 << 1, /* dn_struct_rwlock is locked. */
+	DB_RF_NEVERWAIT		= 1 << 2,
+	DMU_READ_PREFETCH	= 0,	  /* Try speculative prefetch. */
+	DMU_READ_NO_PREFETCH	= 1 << 3, /* Don't prefetch speculatively. */
+	DB_RF_NOPREFETCH	= DMU_READ_NO_PREFETCH,
+	DMU_READ_NO_DECRYPT	= 1 << 4, /* Don't decrypt. */
+	DB_RF_NO_DECRYPT	= DMU_READ_NO_DECRYPT,
+	DMU_DIRECTIO		= 1 << 5, /* Bypass ARC. */
+	DMU_UNCACHEDIO		= 1 << 6, /* Reduce caching. */
+	DMU_PARTIAL_FIRST	= 1 << 7, /* First partial access. */
+	DMU_PARTIAL_MORE	= 1 << 8, /* Following partial access. */
+	DMU_KEEP_CACHING	= 1 << 9, /* Don't affect caching. */
```

Comment:
It took me a while to understand the meaning of the cache flags, and (if I'm right), I think we could name them a bit more clearly. What do you think of:

```
DMU_DIRECTIO     -> DMU_BYPASS_CACHE
DMU_UNCACHEDIO   -> DMU_DONT_CACHE
DMU_KEEP_CACHING -> DMU_RETAIN_CACHE (or `DMU_KEEP_CACHE`)
```

("keep caching" kept making me think of "keep on truckin'", like we're encouraging the cache in these difficult times: "hang in there, cache, you got this". Which says more about my mental state than anything else, maybe).

Reply:
As a naive reader, I might read "direct IO" or "uncached IO" and I wonder if those flags are only for use with those particular IO types or have some special meaning in those cases. On the other hand, more generic terms may mean that you can know less about the overall system to understand them, but might make me harder to understand a broader feature because now you have to know how these pieces fit together.

And yes, I misunderstood the meaning of the last one.

I guess all up, I don't really feel like I love these names, but I don't have any better suggestions, and they probably don't really make it more difficult to understand what's happening. I have asked, and that's the important thing :)